### ISSUE TITLE
spark-avro 3.0.0 read hang

### ISSUE THREAD

**Issue by codingtwinky** [databricks/spark-avro#156]
Hi,

we are working on upgrading to spark 2.0 but we are having some weird behavior that is preventing us from moving on further.  

What is happening is that consistently 51st dataframe read using avro hangs.

This is not intermittent, it is easily reproducible. This is not due to data corruption, I can read troubled data point when i restart. Parquet file read in similar fashion doesn't have this problem.  

We are using s3 and it seems that processing s3 metadata is where it is hung.  Could you possibly guide us on this matter? thanks,

our sbt:
```
scalaVersion := "2.10.5"

crossScalaVersions := Seq("2.10.5", "2.11.4")

// Build dependencies
libraryDependencies ++= Seq(
  "org.apache.spark" %% "spark-core" % "2.0.0" % "provided",
  "org.apache.spark" %% "spark-sql" % "2.0.0" % "provided",
  "org.apache.spark" %% "spark-hive" % "2.0.0" % "provided",
  "org.scalatest" %% "scalatest" % "2.2.4" % "test",
  "org.scalamock" %% "scalamock-scalatest-support" % "3.2" % "test",
  "com.databricks" % "spark-avro_2.10" % "3.0.0"
)

dependencyOverrides ++= Set(
  "com.fasterxml.jackson.core" % "jackson-annotations" % "2.6.5",
  "com.fasterxml.jackson.core" % "jackson-core" % "2.6.5",
  "com.fasterxml.jackson.core" % "jackson-databind" % "2.6.5"
)
```

main thread when hung:
```
java.lang.Thread.sleep(Native Method)
com.amazon.ws.emr.hadoop.fs.shaded.com.amazonaws.http.AmazonHttpClient.doPauseBeforeRetry(AmazonHttpClient.java:1475)
com.amazon.ws.emr.hadoop.fs.shaded.com.amazonaws.http.AmazonHttpClient.pauseBeforeRetry(AmazonHttpClient.java:1439)
com.amazon.ws.emr.hadoop.fs.shaded.com.amazonaws.http.AmazonHttpClient.executeOneRequest(AmazonHttpClient.java:794)
com.amazon.ws.emr.hadoop.fs.shaded.com.amazonaws.http.AmazonHttpClient.executeHelper(AmazonHttpClient.java:607)
com.amazon.ws.emr.hadoop.fs.shaded.com.amazonaws.http.AmazonHttpClient.doExecute(AmazonHttpClient.java:376)
com.amazon.ws.emr.hadoop.fs.shaded.com.amazonaws.http.AmazonHttpClient.executeWithTimer(AmazonHttpClient.java:338)
com.amazon.ws.emr.hadoop.fs.shaded.com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:287)
com.amazon.ws.emr.hadoop.fs.shaded.com.amazonaws.services.s3.AmazonS3Client.invoke(AmazonS3Client.java:3826)
com.amazon.ws.emr.hadoop.fs.shaded.com.amazonaws.services.s3.AmazonS3Client.getObjectMetadata(AmazonS3Client.java:1015)
com.amazon.ws.emr.hadoop.fs.shaded.com.amazonaws.services.s3.AmazonS3Client.getObjectMetadata(AmazonS3Client.java:991)
com.amazon.ws.emr.hadoop.fs.s3n.Jets3tNativeFileSystemStore.retrieveMetadata(Jets3tNativeFileSystemStore.java:212)
sun.reflect.GeneratedMethodAccessor19.invoke(Unknown Source)
sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
java.lang.reflect.Method.invoke(Method.java:498)
org.apache.hadoop.io.retry.RetryInvocationHandler.invokeMethod(RetryInvocationHandler.java:191)
org.apache.hadoop.io.retry.RetryInvocationHandler.invoke(RetryInvocationHandler.java:102)
com.sun.proxy.$Proxy36.retrieveMetadata(Unknown Source)
com.amazon.ws.emr.hadoop.fs.s3n.S3NativeFileSystem.getFileStatus(S3NativeFileSystem.java:780)
org.apache.hadoop.fs.FileSystem.exists(FileSystem.java:1428)
com.amazon.ws.emr.hadoop.fs.EmrFileSystem.exists(EmrFileSystem.java:313)
org.apache.spark.sql.execution.datasources.DataSource.hasMetadata(DataSource.scala:289)
org.apache.spark.sql.execution.datasources.DataSource.resolveRelation(DataSource.scala:324)
org.apache.spark.sql.DataFrameReader.load(DataFrameReader.scala:149)
org.apache.spark.sql.DataFrameReader.load(DataFrameReader.scala:132)
```

**Comment by codingtwinky** [comment]
I'm going crazy over this issue, this simple below code in S3 reproduces my problem.  Does anybody else have this problem as well?

```scala
val avros = (0 to 50).map(n => { 
    spark.read.format("com.databricks.spark.avro").load(s"s3n://avro-files/testFile-${n}")
  }) 
})

// Below code line hangs
    spark.read.format("com.databricks.spark.avro").load(s"s3n://avro-files/testFile-0")
```

**Comment by clockfly** [comment]
It doesn't looks like a Spark issue. You may want o turn on the DEBUG log to see what is failing.

**Comment by codingtwinky** [comment]
Thanks for the suggestion.

I see that connection polling is maxed out and previous connections are persisted.  Trying to figure out how to free these connections or increase them.


Log 51st read log
```
16/08/11 18:12:59 DEBUG PoolingClientConnectionManager: Connection request: [route: {s}->https://<<bucket>>.<<region>>.amazonaws.com][total kept alive: 0; route allocated: 50 of 50; total allocated: 50 of 50]
```

**Comment by zoltanarvai** [comment]
I've encountered the same problem, no solution yet. I'm using multiple dataframes unioned into a single one then I get into this error.

**Comment by codingtwinky** [comment]
This issue is independent from dataframe unions as even non union cases are failing.  This issue seems to be that connections are not released after metadata read.  😿

**Comment by rk101288** [comment]
I am seeing the same issue. Is there a work around for this?

**Comment by codingtwinky** [comment]
Nope

**Comment by st0ckface** [comment]
was seeing the same issue. The workaround we went with was not using this library at all and instead reading our avro files into HadoopRdds (from s3). For example...
```
     val job: Job = Job.getInstance(getSparkContext().hadoopConfiguration)
     FileInputFormat.setInputPaths(job, yourInputPaths)

      val df = sqlContext.createDataFrame(getSparkContext().newAPIHadoopRDD(
        job.getConfiguration,
        classOf[AvroKeyInputFormat[GenericRecord]],
        classOf[AvroKey[GenericRecord]],
        classOf[NullWritable])
       .map(_._1.datum())
       .map(r => (r.get("col1").toString, r.get("col2").toString, r.get("col3").toString)))
       .toDF("col1", "col2", "col3)

**Comment by JoshRosen** [comment]
It looks like the new code never calls `close()` on `DataFileReader`. I think that the right fix here is to use a TaskCompletionCallback to ensure that it's _eventually_ cleaned up a `NextIterator`-like idiom to ensure that it's cleaned up at the end of the input (so things like `union` work correctly).

**Comment by JoshRosen** [comment]
I have a fix for this issue in #174, which I just merged and included in a new 3.0.1 release (complete release notes at https://github.com/databricks/spark-avro/releases/tag/v3.0.1). I've just published this release to Maven Central and Spark Packages, so it should be available soon.

**Comment by nsuguru** [comment]
@JoshRosen : I am getting the same error while using Spark 2.2.0 and pulling data from Hive/S3. I am not using any AVRO files. 
Someone suggested the below 2 lines, which didn't help:
    hc.set("fs.s3a.impl", "org.apache.hadoop.fs.s3a.S3AFileSystem")
    hc.setInt("fs.s3a.connection.maximum", 500)

If I have to close "FsInput" in my spark code, where should I do that?
